### PR TITLE
cocotb: init at 1.2.0

### PR DESCRIPTION
**Pull request by matthuszagh**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by emilazy** [comment]
Thanks for packaging this!

You'll need to add yourself to `maintainers/maintainer-list.nix` too.

The following changes get the tests running and passing:

```diff
diff --git a/pkgs/development/python-modules/cocotb/default.nix b/pkgs/development/python-modules/cocotb/default.nix
index 2da667b77c6..2728da9a3aa 100644
--- a/pkgs/development/python-modules/cocotb/default.nix
+++ b/pkgs/development/python-modules/cocotb/default.nix
@@ -1,19 +1,38 @@
-{ stdenv, buildPythonPackage, fetchPypi, setuptools }:
+{ stdenv, buildPythonPackage, fetchFromGitHub, setuptools, swig, verilog }:
 
 buildPythonPackage rec {
   pname = "cocotb";
   version = "1.2.0";
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "0855gs1wi7k9fwk832czri21wr83gzp0lp2jkaxccyjgbr1d8a8z";
+  src = fetchFromGitHub {
+    owner = pname;
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "091q63jcm87xggqgqi44lw2vjxhl1v4yl0mv2c76hgavb29w4w5y";
   };
 
   propagatedBuildInputs = [
     setuptools
   ];
 
-  doCheck = false;
+  postPatch = ''
+    patchShebangs bin/*.py
+
+    # POSIX portability (TODO: upstream this)
+    for f in \
+      cocotb/share/makefiles/Makefile.* \
+      cocotb/share/makefiles/simulators/Makefile.*
+    do
+      substituteInPlace $f --replace 'shell which' 'shell command -v'
+    done
+  '';
+
+  checkInputs = [ swig verilog ];
+
+  installCheckPhase = ''
+    export PATH=$out/bin:$PATH
+    make test
+  '';
 
   meta = {
     description = "Coroutine based cosimulation library for writing VHDL and Verilog testbenches in Python";
```

**Comment by matthuszagh** [comment]
@emilazy changes made. Thanks for getting the tests working!

**Comment by emilazy** [comment]
I submitted the Makefile changes upstream as https://github.com/cocotb/cocotb/pull/1153.

**Review by emilazy — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -4059,6 +4059,12 @@
     githubId = 427866;
     name = "Matthias Beyer";
   };
+  matthuszagh = {
+    email = "huszaghmatt@gmail.com";
+    github = "matthuszagh";
+    githubId = 7377393;
+    name = "Matt Huszagh";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix]: this should be a separate commit:
```
maintainers: add matthuszagh
pythonPackages.cocotb: init at ...
```

File: pkgs/development/python-modules/cocotb/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      cocotb/share/makefiles/Makefile.* \
+      cocotb/share/makefiles/simulators/Makefile.*
+    do
+      substituteInPlace $f --replace 'shell which' 'shell command -v'
+    done
+  '';
+
+  checkInputs = [ swig verilog ];
+
+  installCheckPhase = ''
```
Comment (jonringer) [inline pkgs/development/python-modules/cocotb/default.nix]: for python packages you want checkPhase, the installCheckPhase get's discarded
```suggestion
  checkPhase = ''
```

File: pkgs/development/python-modules/cocotb/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    export PATH=$out/bin:$PATH
+    make test
+  '';
+
+  meta = {
```
Comment (jonringer) [inline pkgs/development/python-modules/cocotb/default.nix]:
```suggestion
  meta = with stdenv.lib; {
```

File: pkgs/development/python-modules/cocotb/default.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    description = "Coroutine based cosimulation library for writing VHDL and Verilog testbenches in Python";
+    homepage = https://github.com/cocotb/cocotb;
+    license = stdenv.lib.licenses.bsd3;
```
Comment (jonringer) [inline pkgs/development/python-modules/cocotb/default.nix]:
```suggestion
    license = licenses.bsd3;
```

File: pkgs/development/python-modules/cocotb/default.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    maintainers = with stdenv.lib.maintainers; [ matthuszagh ];
```
Comment (jonringer) [inline pkgs/development/python-modules/cocotb/default.nix]:
```suggestion
    maintainers = with maintainers; [ matthuszagh ];
```

**Comment by jonringer** [comment]
please squash all the cocotb related commits into one
```
git reset HEAD~4
git add pkgs/
git commit -m "pythonPackages.cocotb: init at 1.2.0"
```

File: pkgs/development/python-modules/cocotb/default.nix [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    patchShebangs bin/*.py
+
+    # POSIX portability (TODO: upstream this)
+    for f in \
+      cocotb/share/makefiles/Makefile.* \
+      cocotb/share/makefiles/simulators/Makefile.*
+    do
+      substituteInPlace $f --replace 'shell which' 'shell command -v'
+    done
+  '';
```
Comment (jonringer) [inline pkgs/development/python-modules/cocotb/default.nix]: another option could be
```
  nativeBuildInputs = [ which ];
```
this will add which to PATH during build

Comment (emilazy) [reply pkgs/development/python-modules/cocotb/default.nix]: If it was just tests, then I should agree, but this would force every external use of cocotb to run simulations using Nix to include `which` as a dependency, which I think is undesirable (they already have to add an appropriate Verilog/VHDL simulator to the PATH).

Comment (jonringer) [reply pkgs/development/python-modules/cocotb/default.nix]: I'm not familiar with cocotb, and the `shell command -v` is probably the proper thing to do.

But which won't become a runtime dependency, just a build-time one.

Comment (emilazy) [reply pkgs/development/python-modules/cocotb/default.nix]: cocotb installs these Makefiles into `$out` and they're used when doing a simulation workflow with it.

**Review by jonringer — APPROVED** [review]
`nix-review` passes on NixOS
diff LGTM
executable seems to work

```
https://github.com/NixOS/nixpkgs/pull/71288
3 package were build:
python27Packages.cocotb python37Packages.cocotb python38Packages.cocotb
```
```
[nix-shell:/home/jon/.cache/nix-review/pr-71288-1]$ nix path-info -Sh ./results/python37Packages.cocotb
/nix/store/58fll0a5dwag12is7sbjhgz5wlzfh94h-python3.7-cocotb-1.2.0        98.1M
```

**Comment by jonringer** [comment]
@GrahamcOfBorg build python27Packages.cocotb python37Packages.cocotb python38Packages.cocotb

**Comment by jonringer** [comment]
It seems that cocotb support darwin https://cocotb.readthedocs.io/en/latest/quickstart.html#macos-packages , it doesn't look like it's able to resolve gcc correctly, not sure if you all want to pursue mac support, or just mark this linux only

**Comment by matthuszagh** [comment]
I don't know much about osx but happy to take a look if you think it's best to get that working too.

**Comment by jonringer** [comment]
@marsam can you take a look at the failure logs for darwin, it essentially is saying:

```
gcc: command cannot be found
```

I was under the impression that clang is aliased to gcc in most darwin environments.

**Comment by marsam** [comment]
Looks like upstream has a makefile with [gcc harcoded](https://github.com/cocotb/cocotb/blob/master/cocotb/share/makefiles/Makefile.rules#L56-L63). It works with the following:
```diff
diff --git i/pkgs/development/python-modules/cocotb/default.nix w/pkgs/development/python-modules/cocotb/default.nix
index 228ac7904ec..e10ecd8e840 100644
--- i/pkgs/development/python-modules/cocotb/default.nix
+++ w/pkgs/development/python-modules/cocotb/default.nix
@@ -24,6 +24,9 @@ buildPythonPackage rec {
       cocotb/share/makefiles/simulators/Makefile.*
     do
       substituteInPlace $f --replace 'shell which' 'shell command -v'
+      # replace hardcoded gcc. Remove once https://github.com/cocotb/cocotb/pull/1137 gets merged
+      substituteInPlace $f --replace 'gcc' '$(CC)'
+      substituteInPlace $f --replace 'g++' '$(CXX)'
     done
   '';
```

**Review by marsam — APPROVED** [review]

**Comment by cmarqu** [comment]
There is an open PR for cocotb to fix this and maybe more related problems: https://github.com/cocotb/cocotb/pull/1137

**Comment by marsam** [comment]
Thanks for your patience. I added the fix for darwin, I hope you don't mind.
